Improve picasso ui

manuelmeister opened the pull request:
| before |  ![Bildschirmfoto 2022-05-23 um 23 19 50](https://user-images.githubusercontent.com/3001985/169907904-d8f5d2af-fbf5-421b-ab50-976265a1f897.png) |  ![Bildschirmfoto 2022-05-23 um 23 19 40](https://user-images.githubusercontent.com/3001985/169907891-157cd148-73a7-4718-874c-2c5e3b412d96.png) |
| - | ------------- | ------------- |
| after | <img width="1248" alt="Bildschirmfoto 2022-05-31 um 13 29 36" src="https://user-images.githubusercontent.com/3001985/171163365-582c4875-ff8b-4fac-b4e5-a9de7afbdf7a.png">   | <img width="377" alt="Bildschirmfoto 2022-05-31 um 13 29 48" src="https://user-images.githubusercontent.com/3001985/171163406-fcc16227-01c9-4b31-a79f-82f056cbdb55.png">  |

manuelmeister committed:
Make picasso ui more usable on mobile and desktop

manuelmeister committed:
Move print details to details menu. Make editable status more intuitive.

usu commented on frontend/src/components/print/print-nuxt/download.js:
```diff
@@ -0,0 +1,26 @@
+
+const PRINT_SERVER = window.environment.PRINT_SERVER
+
+export async function download (config) {
```
smells like code repetition 👃😄

usu commented on frontend/src/views/camp/CampProgram.vue:
```diff
@@ -6,24 +6,57 @@ Show all activity schedule entries of a single period.
+            </v-list-item-title>
+          </v-list-item>
+          <v-divider />
+          <v-list-item @click="printNuxt">
```
Du musst jetzt hier ziemlich viel print-Login in die `CampProgram.vue` Komponente ziehen. Mach doch eine Komponente analog zu `pdf-download-button-nuxt`, welche als v-list-item rendert anstatt als Button (z.B. pdf-download-menu-item-nuxt). Dort kannst du dir entweder download.js als Methode hineinziehen, oder nutzt ein Mixin.

`mdi-timer-sand` als statisches Icon finde ich nicht ideal, da gefällt mir der Spinner besser. Visuell einfach zu sehen, dass da was am Rendern ist.

Mit dem @click kann ich jetzt das PDF drucken mehrmals triggern (das ist bei VBtn automatisch disabled, wenn der loading-Spinner läuft).

Gibt es eine Möglichkeit, das Menu offen zu lassen, wenn ich auf print klicke?

manuelmeister replied:
Fixed in d038380 and 394f03d

manuelmeister replied:
Würdest du es nur offen lassen, wenn man auf die Downloads klickt?

usu commented on frontend/src/components/program/picasso/Picasso.vue:
```diff
@@ -380,10 +384,16 @@ export default {
 
   @media #{map-get($display-breakpoints, 'xs-only')}{
-      height:calc(100vh - 120px);
+    position: fixed;
```
Der "add activity button" nimmt jetzt in der Mobile-View ziemlich viel Platz weg. Nicht sicher, ob das vorher anders war und ob der Code hier dafür verantwortlich ist oder nicht.

manuelmeister replied:
Da gibt es keinen Unterschied auf meinen Geräten. Wie sieht das denn bei dir auf dev.ecamp3.ch und bei meinem PR auf einem Android aus?

usu commented on frontend/src/views/camp/CampProgram.vue:
```diff
@@ -6,24 +6,57 @@ Show all activity schedule entries of a single period.
+              <v-icon v-if="nuxtIsPrinting">mdi-timer-sand</v-icon>
+              <v-icon v-else>mdi-nuxt</v-icon>
+            </v-list-item-icon>
+            <v-list-item-title>PDF herunterladen</v-list-item-title>
```
Translation missing

BacLuc commented:
removed the deploy label because our cluster could not handle 3 deployments

manuelmeister committed:
Fix picasso height on mobile

manuelmeister committed:
Respect safe-area-inset for placement of interactive elements

manuelmeister committed:
Create dedicated download list items

manuelmeister committed:
Prevent multiple printing orders

manuelmeister committed:
Merge remote-tracking branch 'upstream/devel' into feature/ui-picasso-mobile

# Conflicts:
#	frontend/src/views/camp/NavigationCamp.vue

carlobeltrame commented on frontend/src/components/print/print-nuxt/DownloadNuxtPdfListItem.vue:
```diff
@@ -0,0 +1,77 @@
+      <v-icon v-else>mdi-nuxt</v-icon>
+    </v-list-item-icon>
+    <v-list-item-title>
+      {{ $tc("components.print.localPdfDownloadButton.label") }}
```
Create a new translation key for the new component

manuelmeister replied:
Fixed in fcee4dc9

carlobeltrame commented on frontend/src/components/print/print-react/DownloadReactPdfListItem.vue:
```diff
@@ -0,0 +1,71 @@
+      <v-icon v-else>mdi-react</v-icon>
+    </v-list-item-icon>
+    <v-list-item-title>
+      {{ $tc("components.print.localPdfDownloadButton.label") }}
```
Create a new translation key for the new component

manuelmeister replied:
Fixed in fcee4dc9

carlobeltrame commented on frontend/src/components/print/print-react/DownloadReactPdfListItem.vue:
```diff
@@ -0,0 +1,71 @@
+
+      this.loading = true
+
+      // lazy load generatePdf to avoid loading complete react-pdf when showing PDF download buton
```
```suggestion
      // lazy load generatePdf to avoid loading complete react-pdf when showing PDF download button
```
(typo also exists in the button component)

manuelmeister replied:
Fixed in fcee4dc9

carlobeltrame commented on frontend/src/components/print/print-react/DownloadReactPdfListItem.vue:
```diff
@@ -0,0 +1,71 @@
+          label: this.$tc('components.print.localPdfDownloadButton.error'),
+          trace: error
+        })
+        return
```
You also have to reset loading to false in the error case

manuelmeister replied:
Fixed in fcee4dc9

carlobeltrame commented on frontend/src/components/program/picasso/Picasso.vue:
```diff
@@ -41,10 +41,10 @@ Listing all given activity schedule entries in a calendar view.
           </span> {{ $date.utc(date).format($tc('components.camp.picasso.datetime.date', widthPluralization)) }}
         </div>
         <day-responsibles
+          :class="{ 'ec-daily_head-day-responsibles--readonly': !editable }"
```
Can you document somewhere how to construct these complicated class names? Why is `daily_head` separated with an underscore, but `day-responsibles` with a dash? How do you decide what needs to be in there? And specifically here, why can't we just use scoped styles in the DayResponsibles component?

manuelmeister replied:
Used scoped css 69dcb1b0

carlobeltrame commented on frontend/src/components/program/picasso/Picasso.vue:
```diff
@@ -380,10 +384,24 @@ export default {
+    position: fixed;
+    height: inherit;
+    top: 48px;
+    bottom: calc(56px + env(safe-area-inset-bottom));
```
What is `env(safe-area-inset-bottom)`?

manuelmeister replied:
https://developer.mozilla.org/en-US/docs/Web/CSS/env#values 
Without spacing, I have difficulties to hit the buttons in the bottom navigation.

carlobeltrame replied:
> For rectangular viewports, like your average laptop monitor, their value is equal to zero. For non-rectangular displays — like a round watch face — the four values set by the user agent form a rectangle such that all content inside the rectangle is visible.

Why do you need the `env(...)` stuff at all? Are you optimizing the display for smart watches? :rofl: If not, the value should just be 0, or am I understanding this wrong?

manuelmeister replied:
![Safe area insets](https://felgo.com/wp-content/uploads/2018/05/fullscreen-app-notch-issue-iphonex.png)

It's an Apple problem ;) 
https://webkit.org/blog/7929/designing-websites-for-iphone-x/

carlobeltrame replied:
God damnit Apple

manuelmeister replied:
🤷‍♂️

carlobeltrame commented on frontend/src/views/camp/CampProgram.vue:
```diff
@@ -6,24 +6,45 @@ Show all activity schedule entries of a single period.
+              <v-icon v-else>mdi-lock-open</v-icon>
+            </v-list-item-icon>
+            <v-list-item-title>
+              {{ editMode ? 'Sperren' : 'Entsperren' }}
```
Use translations here

carlobeltrame commented on frontend/src/views/camp/CampProgram.vue:
```diff
@@ -6,24 +6,45 @@ Show all activity schedule entries of a single period.
+          <v-list-item @click="editMode = !editMode">
+            <v-list-item-icon>
+              <v-icon v-if="editMode">mdi-lock</v-icon>
+              <v-icon v-else>mdi-lock-open</v-icon>
```
I find it near impossible to see the difference between these two icons on a desktop screen, at this light grey coloring. Is there a more obvious unlocked icon? Could we help this with different colors / lightness / supporting text / tooltips?

manuelmeister replied:
Fixed in 4b77e6ed

carlobeltrame commented on frontend/src/components/program/picasso/Picasso.vue:
```diff
@@ -435,12 +458,46 @@ export default {
+    overflow: hidden;
+  }
+
+  .ec-daily_head-day-responsibles--readonly .v-input__append-inner {
```
If you start heavily modifying the Vuetify inputs, please make sure that accessibility is not lost. I cannot find a way to select the day responsible via pressing Tab now. If I revert your CSS changes in this file it starts working again, so it probably is something to do with hiding this input.

manuelmeister replied:
Can you try again?

carlobeltrame replied:
Hmm no does not work for me... I can tab-focus the three dots menu, then the next thing is the scrollable picasso area, and then the schedule entries' edit buttons. I tabbed through all elements on the page but never arrived at the dropdowns. Does it work on your machine? I'm using Firefox 100.0.2 on Ubuntu.

manuelmeister replied:
Now it "works". Fixed in 7c7f176. The problem is, v-select readonly behaves strange. (click and keyboard readonly state are switched) ?!

carlobeltrame replied:
Using the keyboard it works now, but now I can only open the select using the mouse when the picasso is locked. When unlocked, clicking the select does nothing. This seems like a Vuetify bug to me which we should be able to raise and possibly fix there...

manuelmeister replied:
I've opened an issue: https://github.com/vuetifyjs/vuetify/issues/15173

manuelmeister replied:
I removed the openOnClick option, this means you can't click the chip to activate the menu. Until this issue is fixed, I think this is a good compromise.

carlobeltrame replied:
Works as expected now, thanks!

carlobeltrame commented on frontend/src/components/program/picasso/DayResponsibles.vue:
```diff
@@ -1,17 +1,18 @@
             chips
-            deletable-chips
             small-chips
+            persistent-placeholder
```
I don't like this to be honest. The text is redundant on every day, and shows up even when not a single day responsible is selected in the whole camp.

manuelmeister replied:
Would it be better for you if the select had no label?

carlobeltrame replied:
I guess so, as long as it is still clear that I can click to open a menu there

manuelmeister replied:
Fixed in 69dcb1b0

carlobeltrame commented on frontend/src/components/program/picasso/DayResponsibles.vue:
```diff
@@ -1,17 +1,18 @@
+  <v-skeleton-loader v-if="isLoading" type="text"
+                     height="46"
+                     class="ma-2" />
   <e-select v-else
```
Is there a way to make just this select automatically close when I select an entry? I would assume there is rarely more than one day responsible. At least on mobile I have a hard time finding a place to tap outside the select in order to close it when I am done selecting. Clicking the dropdown arrow again isn't an option either, because the arrow is covered when the select is open.

manuelmeister replied:
Fixed in 69dcb1b0

carlobeltrame commented on frontend/src/components/program/picasso/Picasso.vue:
```diff
@@ -318,14 +318,18 @@ export default {
+    })
 
     // scroll a bit down to hide the night hours
     const scroller = this.$el.querySelector('.v-calendar-daily__scroll-area')
```
This auto-scrolling logic does not work anymore now on my machine. Fix it or remove it.

manuelmeister replied:
Fixed in e38ac5bb

carlobeltrame reviewed (changes requested):
In the screenshot there is now more empty space below the Picasso on desktop. Why is that?

manuelmeister committed:
Introduce context dependent download pdf label

manuelmeister committed:
Improve day responsible design

manuelmeister committed:
Fix picasso scroll issue

manuelmeister committed:
Make readonly mode more obvious

carlobeltrame commented on docker-compose.yml:
```diff
@@ -6,7 +6,7 @@ services:
     ports:
       - '3000:3000'
-      - '9229:9229' # jest debug 
+      - '9229:9229' # jest debug
```
Please revert your changes to this file. You can also instruct Git to ignore all your changes to this file:
```bash
git update-index --assume-unchanged docker-compose.yml
```
But this does not survive branch switching, and sometimes it's also more confusing than it helps.
To revert this git mind control:
```bash
git update-index --no-assume-unchanged docker-compose.yml
```

manuelmeister replied:
This is quite annoying, because without these changes (I did not want to commit) it doesn't work for me 😕

manuelmeister replied:
Maybe I can rely on the different PhpStorm changelists

carlobeltrame commented on frontend/src/components/print/print-nuxt/DownloadNuxtPdfListItem.vue:
```diff
@@ -0,0 +1,77 @@
+      <v-icon v-else>mdi-nuxt</v-icon>
+    </v-list-item-icon>
+    <v-list-item-title>
+      {{ $tc("components.print.pdfDownload.nuxt.label") }}
```
https://github.com/ecamp/ecamp3/wiki/localization-and-translations#translation-file-structure

The translation keys must match the file structure. This one should be `components.print.printNuxt.downloadNuxtPdfListItem.label`.
This was our agreement when we decided not to use component-local translations. The idea is that this way we can easily find out where a translation is used, delete unused translations and maybe some day even introduce automatic prefixing such that you could just use the translation key `.label` here.

manuelmeister replied:
Okay, I'll change it in the other components as well.

manuelmeister replied:
Nevermind, I'll remove this button anyway soon. All other print components do not abide to those rules.

manuelmeister replied:
Fixed in 27ee19e

manuelmeister committed:
Match translation path to file path

manuelmeister committed:
Fix readonly and input

This still has an issue, because vuetify handles keyboard and click differently

manuelmeister committed:
Move period switcher permanently into header of picasso

manuelmeister commented:
I just made a review with someone from UX and realized that it would be best if we used the period dropdown for the desktop picasso as well. The reasons are that it is the same on mobile and the change of the period is not frequent.

manuelmeister committed:
Remove openOnClick

This can be reverted, when https://github.com/vuetifyjs/vuetify/issues/15173 is resolved

BacLuc commented on frontend/src/components/program/picasso/DayResponsibles.vue:
```diff
@@ -33,6 +42,11 @@ export default {
       required: true
+    },
+
+    readonly: {
```
minor:
i called this disabled in other components (had to to this a lot to disable some things for guests) to be consistent with the disabled attribute from html.
I know disabled attribute has inverse logic, but i decided to follow the html way.
And i would keep it consistent.

But MaterialItem it's readonly, and in BasePicker we even have both...?

manuelmeister replied:
I would rather have it as readonly, because disabled makes it for a11y impossible to focus and to realize that something is even there.

BacLuc commented on frontend/src/components/program/picasso/DayResponsibles.vue:
```diff
@@ -129,4 +143,13 @@ export default {
     padding-right: 0 !important;
   }
+
+  ::v-deep .v-input {
```
as info:
Make a rule that targets child components.
v-deep
Note: usage without arguments is deprecated since Vue 3.0.
Deprecated

manuelmeister replied:
Can I do this already in v2?

BacLuc commented on frontend/src/components/program/picasso/Picasso.vue:
```diff
@@ -43,8 +43,7 @@ Listing all given activity schedule entries in a calendar view.
           :period="period"
-          :disabled="!editable"
-          class="mt-1 mx-2" />
+          :readonly="!editable" />
```
ok, how did this even work before?
Why did you change it to readonly?

manuelmeister replied:
https://github.com/ecamp/ecamp3/pull/2744#discussion_r895220699

BacLuc commented on frontend/src/components/program/picasso/Picasso.vue:
```diff
@@ -459,13 +515,69 @@ export default {
 }
 
+.readonlyEntry .v-event-title {
+  color: white;
```
now the text is white (which looks good, maybe we have to disable light colors for categories), but the underline is blue.

Comes from:

.v-application a {
 color: #fff;
}

manuelmeister replied:
This is on purpose, to indicate a link. I can disable the underline, but I think it helps to indicate, that you have different affordances between readonly/link mode and editing mode

BacLuc commented on frontend/src/views/camp/CampProgram.vue:
```diff
@@ -42,6 +62,17 @@ Show all activity schedule entries of a single period.
         </template>
       </template>
     </schedule-entries>
+    <v-snackbar v-model="showError" app :timeout="10000">
```
does this clear the showError flag if the timeout passes?

manuelmeister replied:
yes, this has just moved and was in the button component before

BacLuc reviewed (approved):
Very cool, thank you

usu commented on frontend/src/components/print/print-nuxt/DownloadNuxtPdfListItem.vue:
```diff
@@ -0,0 +1,77 @@
+<template>
+  <v-list-item @click="generatePdf">
```
```suggestion
  <v-list-item @click="generatePdf" :disabled="loading">
```

manuelmeister replied:
Perfect, here I think it makes sense to disable it completely

usu commented on frontend/src/components/print/print-react/DownloadReactPdfListItem.vue:
```diff
@@ -0,0 +1,72 @@
+<template>
+  <v-list-item @click="generatePdf">
```
```suggestion
  <v-list-item @click="generatePdf" :disabled="loading">
```

usu commented on frontend/src/components/print/print-nuxt/DownloadNuxtPdfListItem.vue:
```diff
@@ -0,0 +1,77 @@
+const PRINT_SERVER = window.environment.PRINT_SERVER
+
+export default {
+  name: 'DownloadNuxtPdfListItem',
```
Would be good to keep the naming consistent.

If we call this `DownloadNuxtPdfListItem`, then the Button should be called `DownloadNuxtPdfButton`

manuelmeister replied:
Should I rename it, even though I'll remove the button in next PR targeting the activity.

usu commented on frontend/src/views/camp/CampProgram.vue:
```diff
@@ -70,7 +101,10 @@ export default {
+      showError: null,
+      error: null,
+      nuxtIsPrinting: false,
+      reactIsPrinting: false
```
I think both `nuxtIsPrinting` and `reactIsPrinting` are not needed anymore

usu reviewed (dismissed):
Visually looks good to me. A few minor comments. Other than that, ready to merge from my side

usu commented on frontend/src/components/program/PeriodSwitcher.vue:
```diff
@@ -4,20 +4,28 @@
+          <v-list-item-content>
+            <v-list-item-title>{{ item.description }}</v-list-item-title>
+            <v-list-item-subtitle>
+              {{ $date.utc(item.start).format($tc('global.datetime.dateShort')) }} - {{ $date.utc(item.end).format($tc('global.datetime.dateShort')) }}
```
Alternatively, you could also use `rangeShort(item.start, item.end)` from `dateHelperUTCFormatted.js`

manuelmeister replied:
I don't think this makes sense, as this includes the time. I've created a new dateRange helper.

usu replied:
Ah yes, true. Mixed it up. Well solved.

manuelmeister committed:
Merge remote-tracking branch 'upstream/devel' into feature/ui-picasso-mobile

manuelmeister committed:
Remove unused printing state variables

manuelmeister committed:
Disable item when generating pdf

manuelmeister committed:
Implement dateRange for period durations

manuelmeister committed:
Rename Download Buttons to match Download Listitem
